[April 16] ERD and entity analysis for game session entities (#85)
rightoneducation/righton-app: docs/mobile/game-session/game_session_db_design.md

# DynamoDB Design

## Entities

### Config

This entity contains the global settings and configurable parts of the application. The list of items included here is as follow

> - Points granted to each team estimated the right popular trick answer
> - Points granted per to teams guessed the right answer

Reviewer: ```suggestion
- Points granted to each team for guessing the most popular trick answer
- Points granted to teams for guessing the right answer
```

### GameSession

The higher level entity contains all information related to a game from start to finish.

- Game: A game object that comes from Aurora. It's a snapshot of the data when the game starts
- Teams: An array of team entities that involved in the game
- StartedDateTimeUTC: Date time of when the game started in UTC

> - MultiDeviceMode: Boolean indicating if it's a multi-device with facilitator or not

Reviewer 2: Maybe: "Boolean indicating multi-device scenario" (if True, then one device among a team's multiple devices will be randomly assigned the Facilitator; if False, then there is only one device per team and that device will also have Facilitator privileges)

Reviewer 2: Variables configured on the teacher side that affect student game sessions include (variable names are just draft suggestions):
- VotingTime: Amount of time that voting teams have to vote (this equals the amount of time that the team coming up with the trick answers (aka QuestionTeam?) has to come up with and select their trick answers.
- HintsRevealTime: Amount of time that passes before hints are revealed. This must be less than VotingTime.
- We should set default values for both these variables and allow teachers to adjust them before a game starts. VotingTime should be something that a teacher can override in the middle of game (e.g., if a teacher sees that all teams are ready to proceed, the teacher should be able to move on to the results screens without having to wait for the entire VotingTime to have passed).

Author: Good point, per our discussion, I added `VotingTime` and `HintsRevealTime` to `Config` also and also here. Regarding `VotingTime` modification at runtime, we should probably just do this through UI so teachers can end it by just tapping a button.

### Team

- TeamNo / TeamId: Numeric unique number assigned to the team
- Team name: Name chosen to show fo the team. This helps for future substituting `team 1` string with chosen team names
- Question: The chosen question for the team

> - GameCode: A 4-digit code to enter to join the team

Reviewer: We use codes to join teams? I thought there was one code for the whole game.

Reviewer 2: Though "No" is indeed a common abbreviation for Number, might TeamNum or something along those lines be worth considering?

Author: We talked about it, the idea here was assigning codes for each team so audience joins with a code that assigned them to a specific team. For now, that's not needed so moving this to `GameSession`

### TeamMember

- TeamId: The teamId this member is a member of
- DeviceId: Unique number associated to the device
- IsFacilitator: A boolean indicating if this device is the facilitator in multi-device mode

> ### Question

Reviewer: These Question entities would also be in the GameSession.Game. Is this section describing the field types for Questions, or are we thinking about pulling them out of the Game object? If so, does the data ever change?

Author: Yes, actually I missed to layout the `Game` entity here. I'm listing all entities that are involved on the mobile side and then their relationships. Then we can decide if questions will be `GameSession.Game.Questions` or be accessible as a separate entity.

- QuestionId: Unique number for each question
- Question: The text for the question
- Hints: an array of hints for the question
- ImageURL: URL of the image associated to this question
- Answer: The right answer for this question

### TeamAnswer

- AnswerId: A unique id for each answer

> - Text: The answer chosen by a team member

Reviewer: How do I compare this to the related question? Should we add a reference to the question here?

- TeamMember: DeviceId entered the answer
- IsTrickAnswer: Indicates if this answer is chosen to be a trick answer

### TeamRoundResults

- TeamId: The unique number identifies the team choosing the answer

> - AnswerId: Identifies which answer

Reviewer: is this the answer they gave that was correct?

> - QuestionTeamID: Identifies which team's question is this

Reviewer 2: Might it make sense to further differentiate TeamId from QuestionTeamId -- e.g., perhaps updating the former to AnswerTeamId? For consistency, perhaps QuestionTeamID might be tweaked to QuestionTeamId.

Reviewer 3: Following up with Sinclair's comment, it may make consistent sense to update AnswerId to AnswerTeamId as well as the formers if we're looking to distinguish them. Otherwise, I think it makes sense design wise.

- Title: the title of the game
- Description: The description set for this game

### Team

- TeamId: Numeric unique number assigned to the team
- GameSessionId: Id of the game session that this team belongs to
- Name: Name chosen to show for the team. This helps for future substituting `team 1` string with chosen team names
- Question: The chosen question for the team
- TrickiestAnswerId: Indicates the answer the team has chosen to be the most tricky one

### Question

These information are coming as part of Game entity.

- QuestionId: Unique number for each question
- Question: The text for the question
- Hints: an array of hints for the question
- ImageURL: URL of the image associated to this question
- Answer: The right answer for this question

### TeamMember

- TeamId: The teamId this member is a member of
- DeviceId: Unique number associated to the device
- IsFacilitator: A boolean indicating if this device is the facilitator in multi-device mode

### TeamAnswer

- AnswerId: A unique id for each answer
- QuestionId: The identifier for the question this answer is for
- Text: The answer chosen by a team member
- DeviceId: DeviceId entered the answer

> - IsChosen: Indicates if this answer is chosen to be a trick answer

Author: Change this to `IsChosenTrickAnswer`
